File: frontend/src/services/menu.ts
/**
 * Menu API service.
 * 
 * Handles all menu item-related API calls.
 */

import apiClient, { getErrorMessage } from './api';
import {
  MenuItem,
  IngredientTag,
  PrepType,
  CreateMenuItemData,
  UpdateMenuItemData,
} from '../types/menu';

/**
 * API response structure for menu endpoints.
 */
interface MenuAPIResponse {
  success: boolean;
  data: MenuItem | MenuItem[] | string[];
  message: string;
  timestamp: string;
}

/**
 * Normalize a menu item so it always has `id` and `price` (backend may send `_id`/`id`, `price`/`price_amount`).
 * Price is stored in paise; ensure it's a number for display and totals.
 */
function normalizeMenuItem(item: Record<string, unknown>): MenuItem {
  const id = (item.id as string) ?? (item._id as string) ?? '';
  const rawPrice = item.price ?? item.price_amount;
  let price = 0;
  if (typeof rawPrice === 'number' && !Number.isNaN(rawPrice)) {
    price = Math.max(0, Math.round(rawPrice));
  } else if (typeof rawPrice === 'string') {
    const parsed = parseFloat(rawPrice);
    price = Number.isNaN(parsed) ? 0 : Math.max(0, Math.round(parsed));
  }
  return { ...item, id, price } as MenuItem;
}

/**
 * Extract array from API response body (handles both { data: [] } and raw array).
 */
function extractDataArray<T>(body: unknown): T[] {
  if (Array.isArray(body)) return body as T[];
  const data = (body as { data?: unknown })?.data;
  return Array.isArray(data) ? (data as T[]) : [];
}

/**
 * Get all menu items for the authenticated user's restaurant.
 * 
 * @param includeInactive - Whether to include inactive items
 * @param category - Optional category filter
 * @returns Promise resolving to array of menu items
 * @throws Error if request fails
 */
export async function getMenuItems(
  includeInactive = false,
  category?: string
): Promise<MenuItem[]> {
  try {
    const params: Record<string, string | boolean> = {
      include_inactive: includeInactive,
    };
    
    if (category) {
      params.category = category;
    }
    
    const response = await apiClient.get<MenuAPIResponse>('/menu/items', {
      params,
    });
    
    const raw = response.data;
    const list = extractDataArray<Record<string, unknown>>(raw);
    return list.map((item) => normalizeMenuItem(item));
  } catch (error) {
    throw new Error(getErrorMessage(error));
  }
}

/**
 * Get all unique menu categories.
 * 
 * @returns Promise resolving to array of category names
 * @throws Error if request fails
 */
export async function getCategories(): Promise<string[]> {
  try {
    const response = await apiClient.get<MenuAPIResponse>('/menu/items/categories');
    return extractDataArray<string>(response.data);
  } catch (error) {
    throw new Error(getErrorMessage(error));
  }
}

/**
 * Get a specific menu item by ID.
 * 
 * @param itemId - Menu item database ID
 * @returns Promise resolving to menu item data
 * @throws Error if item not found or request fails
 */
export async function getMenuItem(itemId: string): Promise<MenuItem> {
  try {
    const response = await apiClient.get<MenuAPIResponse>(`/menu/items/${itemId}`);
    const raw = (response.data as { data?: Record<string, unknown> })?.data;
    if (!raw || typeof raw !== 'object') {
      throw new Error('Invalid menu item response');
    }
    return normalizeMenuItem(raw);
  } catch (error) {
    throw new Error(getErrorMessage(error));
  }
}

/**
 * Create a new menu item.
 * 
 * @param data - Menu item creation data
 * @returns Promise resolving to created menu item
 * @throws Error if creation fails
 */
export async function createMenuItem(data: CreateMenuItemData): Promise<MenuItem> {
  try {
    const response = await apiClient.post<MenuAPIResponse>('/menu/items', data);
    const raw = (response.data as { data?: Record<string, unknown> })?.data;
    if (!raw || typeof raw !== 'object') {
      throw new Error('Invalid menu item response');
    }
    return normalizeMenuItem(raw);
  } catch (error) {
    throw new Error(getErrorMessage(error));
  }
}

/**
 * Update menu item details.
 * 
 * @param itemId - Menu item database ID
 * @param data - Fields to update
 * @returns Promise resolving to updated menu item
 * @throws Error if update fails
 */
export async function updateMenuItem(
  itemId: string,
  data: UpdateMenuItemData
): Promise<MenuItem> {
  try {
    const response = await apiClient.put<MenuAPIResponse>(`/menu/items/${itemId}`, data);
    const raw = (response.data as { data?: Record<string, unknown> })?.data;
    if (!raw || typeof raw !== 'object') {
      throw new Error('Invalid menu item response');
    }
    return normalizeMenuItem(raw);
  } catch (error) {
    throw new Error(getErrorMessage(error));
  }
}

/**
 * Delete a menu item (soft delete).
 * 
 * @param itemId - Menu item database ID
 * @returns Promise resolving when deletion is complete
 * @throws Error if deletion fails
 */
export async function deleteMenuItem(itemId: string): Promise<void> {
  try {
    await apiClient.delete(`/menu/items/${itemId}`);
  } catch (error) {
    throw new Error(getErrorMessage(error));
  }
}

File: backend/app/api/v1/menu.py
"""
Menu API endpoints.

Handles menu item management operations.
"""

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from motor.motor_asyncio import AsyncIOMotorDatabase

from app.core.database import get_database
from app.core.dependencies import get_current_user
from app.models.menu_item import (
    MenuItemCreate,
    MenuItemResponse,
    MenuItemUpdate,
)
from app.models.user import UserResponse
from app.repositories.menu_repository import MenuRepository
from app.utils.response import success_response, error_response

router = APIRouter(prefix="/menu/items", tags=["Menu"])


def get_menu_repository(db: AsyncIOMotorDatabase = Depends(get_database)) -> MenuRepository:
    """Dependency to get menu repository instance."""
    return MenuRepository(db)


@router.post("", response_model=dict, status_code=status.HTTP_201_CREATED)
async def create_menu_item(
    menu_item: MenuItemCreate,
    current_user: UserResponse = Depends(get_current_user),
    repo: MenuRepository = Depends(get_menu_repository)
):
    """
    Create a new menu item.
    
    Args:
        menu_item: Menu item creation data.
        current_user: Current authenticated user (auto-injected).
        repo: Menu repository instance.
        
    Returns:
        Created menu item data.
        
    Raises:
        HTTPException: If creation fails.
    """
    try:
        created_item = await repo.create(menu_item)
        
        return success_response(
            data=MenuItemResponse.model_validate(created_item).model_dump(),
            message="Menu item created successfully"
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=error_response(
                code="CREATION_FAILED",
                message=f"Failed to create menu item: {str(e)}"
            )
        )


@router.get("", response_model=dict)
async def get_menu_items(
    include_inactive: bool = Query(False, description="Include inactive menu items"),
    category: Optional[str] = Query(None, description="Filter by category"),
    current_user: UserResponse = Depends(get_current_user),
    repo: MenuRepository = Depends(get_menu_repository)
):
    """
    Get all menu items.
    
    Args:
        include_inactive: Whether to include inactive items.
        category: Optional category filter.
        repo: Menu repository instance.
        
    Returns:
        List of menu items.
    """
    items = await repo.get_all(include_inactive=include_inactive, category=category)
    
    return success_response(
        data=[MenuItemResponse.model_validate(item).model_dump() for item in items],
        message=f"Retrieved {len(items)} menu item(s)"
    )


@router.get("/categories", response_model=dict)
async def get_categories(
    current_user: UserResponse = Depends(get_current_user),
    repo: MenuRepository = Depends(get_menu_repository)
):
    """
    Get all unique menu categories.
    
    Args:
        repo: Menu repository instance.
        
    Returns:
        List of category names.
    """
    categories = await repo.get_categories()
    
    return success_response(
        data=categories,
        message=f"Retrieved {len(categories)} category(ies)"
    )


@router.get("/{item_id}", response_model=dict)
async def get_menu_item(
    item_id: str,
    current_user: UserResponse = Depends(get_current_user),
    repo: MenuRepository = Depends(get_menu_repository)
):
    """
    Get a specific menu item by ID.
    
    Args:
        item_id: Menu item database ID.
        repo: Menu repository instance.
        
    Returns:
        Menu item data.
        
    Raises:
        HTTPException: If menu item not found.
    """
    item = await repo.get_by_id(item_id)
    
    if not item:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=error_response(
                code="NOT_FOUND",
                message="Menu item not found"
            )
        )
    
    return success_response(
        data=MenuItemResponse.model_validate(item).model_dump(),
        message="Menu item retrieved successfully"
    )


@router.put("/{item_id}", response_model=dict)
async def update_menu_item(
    item_id: str,
    update_data: MenuItemUpdate,
    current_user: UserResponse = Depends(get_current_user),
    repo: MenuRepository = Depends(get_menu_repository)
):
    """
    Update menu item details.
    
    Args:
        item_id: Menu item database ID.
        update_data: Fields to update.
        current_user: Current authenticated user (auto-injected).
        repo: Menu repository instance.
        
    Returns:
        Updated menu item data.
        
    Raises:
        HTTPException: If menu item not found.
    """
    updated_item = await repo.update(item_id, update_data)
    
    if not updated_item:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=error_response(
                code="NOT_FOUND",
                message="Menu item not found"
            )
        )
    
    return success_response(
        data=MenuItemResponse.model_validate(updated_item).model_dump(),
        message="Menu item updated successfully"
    )


@router.delete("/{item_id}", response_model=dict)
async def delete_menu_item(
    item_id: str,
    current_user: UserResponse = Depends(get_current_user),
    repo: MenuRepository = Depends(get_menu_repository)
):
    """
    Soft delete a menu item (mark as inactive).
    
    Args:
        item_id: Menu item database ID.
        current_user: Current authenticated user (auto-injected).
        repo: Menu repository instance.
        
    Returns:
        Success message.
        
    Raises:
        HTTPException: If menu item not found or deletion fails.
    """
    success = await repo.soft_delete(item_id)
    
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=error_response(
                code="NOT_FOUND",
                message="Menu item not found"
            )
        )
    
    return success_response(
        data={"id": item_id},
        message="Menu item deleted successfully"
    )
